indico/modules/events/registration/models/items.py

```diff
@@ -177,6 +177,11 @@ class RegistrationFormItem(db.Model):
         db.CheckConstraint('current_data_id IS NULL OR type IN ({t.field}, {t.field_pd})'
                            .format(t=RegistrationFormItemType),
                            name='current_data_id_only_field'),
+        db.CheckConstraint('retention_period IS NULL OR '
+                           '((type = {t.field}) OR ((type = {t.field_pd}) AND (personal_data_type IN '
+                           '({pt.affiliation}, {pt.title}, {pt.address}, {pt.phone}, {pt.country}, {pt.position}))))'
```

Comment:
I think you have some unnecessary parentheses ;) And I'd rather use `NOT IN` for the required personal data fields, ie something like this:

```
retention_period IS NULL OR
type = field OR
(type = field_pd AND personal_data_type NOT IN (<required fields>))
```

FWIW i would not hardcode the required fields but get them from the enum using `is_required`